### ISSUE TITLE
[bug] CMakeDeps fails when building fmt as header-only

### ISSUE THREAD

**Issue by jwillikers** [conan-io/conan#9802]
### Environment Details (include every applicable attribute)
  * Operating System+version: Fedora Silverblue 35
  * Compiler+version: GCC 11
  * Conan version: 1.41.0
  * Python version: 3.10

### Steps to reproduce (Include if Applicable)

Attempt to create a package for a header-only library using the CMakeDeps generator.
Just try adding `CMakeDeps` to the generators list for the `fmt` package and attempt to build it with the option `header_only` enabled.
Since it doesn't have a build_type in that situation, Conan raises an exception.

### Logs (Executed commands with output) (Include/Attach if Applicable)

```
Traceback (most recent call last):
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/generators/__init__.py", line 164, in write_generators
    generator = generator_class(conanfile)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conan/tools/cmake/cmakedeps/cmakedeps.py", line 25, in __init__
    self.configuration = str(self._conanfile.settings.build_type)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/model/settings.py", line 266, in __getattr__
    self._check_field(field)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/model/settings.py", line 262, in _check_field
    raise undefined_field(self._name, field, self.fields, self._parent_value)
conans.errors.ConanException: 'settings.build_type' doesn't exist
'settings' possible configurations are none

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/home/jordan/Projects/Ag Leader/fmt-recipe/build.py", line 166, in <module>
    builder.run(base_profile_name=host_profile, base_profile_build_name=build_profile)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/cpt/packager.py", line 585, in run
    self.run_builds(base_profile_name=base_profile_name,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/cpt/packager.py", line 698, in run_builds
    r.run()
  File "/var/home/jordan/.local/lib/python3.10/site-packages/cpt/runner.py", line 145, in run
    self._results = self._conan_api.create(self._conanfile, name=name, version=version,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/conan_api.py", line 93, in wrapper
    return f(api, *args, **kwargs)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/conan_api.py", line 388, in create
    create(self.app, ref, graph_info, remotes, update, build_modes,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/cmd/create.py", line 72, in create
    install_build_and_test(app, test_conanfile_path,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/cmd/test.py", line 26, in install_build_and_test
    deps_install(app=app,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/manager.py", line 83, in deps_install
    installer.install(deps_graph, remotes, build_modes, update, profile_host, profile_build,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 316, in install
    self._build(nodes_by_level, keep_build, root_node, profile_host, profile_build,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 451, in _build
    self._handle_node_cache(node, keep_build, processed_package_refs, remotes)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 541, in _handle_node_cache
    pref = self._build_package(node, output, keep_build, remotes)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 580, in _build_package
    pref = builder.build_package(node, keep_build, self._recorder, remotes)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 239, in build_package
    raise exc
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 226, in build_package
    self._build(conanfile, pref)
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/installer.py", line 133, in _build
    self._generator_manager.write_generators(conanfile, conanfile.build_folder,
  File "/var/home/jordan/.local/lib/python3.10/site-packages/conans/client/generators/__init__.py", line 171, in write_generators
    raise ConanException("Error in generator '{}': {}".format(generator_name,
conans.errors.ConanException: Error in generator 'CMakeDeps': 'settings.build_type' doesn't exist
'settings' possible configurations are none
```

**Comment by memsharded** [comment]
Hi @jwillikers 

Thanks for reporting. Two things here. First, for sure the exception shouldn't be raised that way, and it should be controlled with a more informative message. The second thing is that it seems indeed necessary to have ``build_type`` defined for using this generator:

- If it is a header-only pure recipe, without settings, then the ``build()`` method should be empty, no build to do, and then no generators should be used. Just a ``package()`` method to package those headers.
- If the final package is just a single one, but we want to be able to build and run some tests, for different configurations, then we need that configuration defined, including the ``build_type``. So the recipe should declare the typical ``settings``, and then, only at the ``package_id()`` method, clear them with ``self.info.header_only()``. That would avoid the error completely, and more errors down the road that would happen at build time for not having things fully defined.

**Comment by jwillikers** [comment]
@memsharded Thanks, I'll fix that in the `fmt` package on ConanCenter.
The settings are cleared in the `configure` stage there which causes the problem.
Sorry about that.

```python
def configure(self):
    if self.options.header_only:
        self.settings.clear()
```

**Comment by memsharded** [comment]
yes, exactly, that is quite legacy code, if you want to build something, you cannot clear those settings in ``configure()`` which runs before generate() and build(), that should be changed in recipes. 

Packages in ConanCenter haven't started to use the new integrations yet, because the Conan version is like one release delayed there. But they will be accepted when possible to run them in the ConanCenter CI.

**Comment by jwillikers** [comment]
Closing, fixed in conan-io/conan-center-index#7686.
